qiskit_nature/properties/property.py

```diff
@@ -0,0 +1,99 @@
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The Property base class."""
+
+from abc import ABC, abstractmethod, abstractclassmethod
+from typing import List, Union, Type
+
+from qiskit_nature import QiskitNatureError
+from qiskit_nature.drivers.second_quantization import QMolecule, WatsonHamiltonian
+from qiskit_nature.operators.second_quantization import SecondQuantizedOp
+from qiskit_nature.results import EigenstateResult
+
+
+class Property(ABC):
+    """The Property base class.
+
+    A Property in Qiskit Nature provides the means to give meaning to a given set of raw data.
+    As such, it provides the logic to transform a raw data (as e.g. produced by a
+    `qiskit_nature.drivers.BaseDriver`) into a
+    `qiskit_nature.operators.second_quantization.SecondQuantizedOp`.
+    As such, every Property is an object which constructs an operator to be evaluated during the
+    problem solution and the interface provides the means for a user to write any custom Property
+    (i.e. the user can evaluate custom _observables_ by writing a class which can generate an
+    operator out of a given set of raw data).
+    """
+
+    def __init__(self, name: str) -> None:
+        """
+        Args:
+            name: the name of the property.
+        """
+        self._name = name
+
+    @property
+    def name(self) -> str:
+        """Returns the name."""
+        return self._name
+
+    @name.setter
+    def name(self, name: str) -> None:
+        """Sets the name."""
+        self._name = name
+
+    @abstractclassmethod
+    def from_driver_result(cls, result: Union[QMolecule, WatsonHamiltonian]) -> "Property":
+        """Construct a Property instance from a driver result.
+
+        This method should implement the logic which is required to extract the raw data for a
+        certain property from the result produced by a driver.
+
+        Args:
+            result: the driver result from which to extract the raw data.
+
+        Returns:
+            An instance of this property.
+
+        Raises:
+            QiskitNatureError: if an invalid driver result type is passed.
+        """
+
+    @classmethod
+    def _validate_input_type(
+        cls,
+        result: Union[QMolecule, WatsonHamiltonian],
+        valid_type: Type[Union[QMolecule, WatsonHamiltonian]],
+    ) -> None:
+        if not isinstance(result, valid_type):
+            raise QiskitNatureError(
+                f"You cannot construct an {cls.__name__} from a {result.__class__.__name__}. "
+                f"Please provide a {valid_type.__name__} object instead."
+            )
```

Comment:
Okay so the errors of the CI are actually related to this method. I tried to extract the `isinstance` checks into the `Property` baseclass by doing the above. Essentially, each property can now call something like the following: `cls._validate_input_type(result, QMolecule)` at the beginning of their `from_driver_result` method. This method will check the type of `result` against the provided `valid_type` and raise an error if these don't match.

This is now a problem due to our current duplicate `QMolecule` situation in `qiskit_nature.drivers` (deprecated) and `qiskit_nature.drivers.second_quantization`. Since the properties all use the new location `QMolecule` object, all `isinstance` checks will fail when given an instance of the deprecated object. This happens in the unittests for `tests/drivers/pyscfd` (which test the deprecated code). This also happens for the _stable_ tutorials for the same reason.

Now, I am not entirely sure what the best approach is to resolve this issue. @manoelmarques @woodsp-ibm do you have some suggestions?

Reply:
I guess one approach would be to also import the deprecated `QMolecule` and hard-code a relation between it and it's new version into the `isinstance` check. Something like:
```
elif valid_type == QMolecule and not isinstance(result, LegacyQMolecule):
    # also raise an error
```
The same would have to be done for `WatsonHamiltonian`.

The problem with this approach is that it will only work as long as both `QMolecule` objects are actually still compatible. If this is subject to change before the release, the `from_driver_result` implementations would have to deal with that (maybe we would need a `from_legacy_driver_result`). This can happen (e.g.) if we replace the new `QMolecule` with an `ElectronicDriverResult` (which may not need such a method if it already consists of `Property` objects). Then, the `from_driver_result` method would essentially become a wrapper for legacy objects anyways.